src/_pytest/python.py

```diff
@@ -1724,6 +1726,17 @@ def write_docstring(tw: TerminalWriter, doc: str, indent: str = "    ") -> None:
         tw.line(indent + line)
 
 
+class DeprecatingFuncArgs(Dict[str, object]):
+    def __init__(self, initialnames: Sequence[str]) -> None:
+        super().__init__()
+        self.initialnames: Final = initialnames
+
+    def __getitem__(self, key: str) -> object:
+        if key not in self.initialnames:
```

Comment:
How is to add a flag attribute to `DeprecatingFuncArgs` in order to raise the warning once hence looking `self.initialnames` up only once? Like what it is like now.

Reply:
Haven't looked at the code yet, but from the description: for good-behaving test suites (hopefully the majority) the warning should never be emitted, in which case it will still be costly. So I don't think it's a good solution for the performance issue (though only warning once might be a good idea regardless).